File: src/prototype/transport.m
clc
clear all
close all

globals; 

%% Initialisation %%

N = 1;
Q = 1;

% Matrice de l'opérateur b %
Bm = zeros(2*(Q+1),(N+2)*(Q+1));
Bm(1:Q+1,1:Q+1) = eye(Q+1);
Bm(Q+2:end,end-Q:end) = eye(Q+1);

Bf = [];
for i = 1:N+1
    Bf = blkdiag(Bf,[1 zeros(1,Q+1);zeros(1,Q+1) 1]);
end

B = blkdiag(Bm,Bf);

% matrice d'interpolation %
Im = zeros((N+1)*(Q+1),(N+2)*(Q+1));
for i = 1:(N+1)*(Q+1)
    for j = 1:(N+2)*(Q+1)
        if i == j 
            Im(i,j) = 1;
        elseif j == i+Q+1
            Im(i,j) = 1;
        end
    end
end

dia = zeros(Q+1,Q+2);
for i = 1:Q+1
    for j = 1:Q+2
        if i == j 
            dia(i,j) = 1;
        elseif j == i+1
            dia(i,j) = 1;
        end
    end
end
If = [];
for i = 1:N+1
    If = blkdiag(If,dia);
end

Interp = 0.5*blkdiag(Im,If);

% matrice de projection sur G2 %
pG2 = inv(eye((N+1)*(Q+2)+(N+2)*(Q+1)) + Interp'*Interp);

% Matrice de la divergence %
Dm = zeros((N+1)*(Q+1),(N+2)*(Q+1));
for i = 1:(N+1)*(Q+1)
    for j = 1:(N+2)*(Q+1)
        if i == j 
            Dm(i,j) = -1;
        elseif j == i+Q+1
            Dm(i,j) = 1;
        end
    end
end
dia = zeros(Q+1,Q+2);
for i = 1:Q+1
    for j = 1:Q+2
        if i == j 
            dia(i,j) = -1;
        elseif j == i+1
            dia(i,j) = 1;
        end
    end
end
Df = [];
for i = 1:N+1
    Df = blkdiag(Df,dia);
end

D = [N*Dm Q*Df];

% matrices projection sur C %
A = [D ; B]; 
delta = A*A'; 

normalise = @(f) f/sum(f(:));

sigma = 0.005; mini = 0.000001;
f0 = normalise(mini + gauss(0.1,sigma,N)); 
f1 = normalise(mini + gauss(0.9,sigma,N));% + gauss(0.8,sigma,N) + gauss(0.4,sigma,N)); 

y = [zeros((N+1)*(Q+1),1) ; zeros(2*(Q+1),1) ; reshape([f1;f0],2*(N+1),1)];
Cst = A'*(delta\y);

P = eye((N+1)*(Q+2)+(N+2)*(Q+1)) - A'*(delta\A);

%% Fin Initialisation %%

m = zeros(Q+1,N+1);
f = zeros(Q+1,N+1);
mbar = zeros(Q+1,N+2);
fbar = zeros(Q+2,N+1);

V = [reshape(m,(N+1)*(Q+1),1);reshape(f,(N+1)*(Q+1),1)];
U = [reshape(mbar,(N+2)*(Q+1),1);reshape(fbar,(N+1)*(Q+2),1)];

alpha = 1.998; g = 0.0125;

wU0 = zeros(size(U)); wV0 = zeros(size(V));
zU0 = zeros(size(U)); zV0 = zeros(size(V));

% % test autre initialisation % 
% fbar = zeros(Q+2,N+1);
% t = repmat(linspace(1,0,Q+2)',1,N+1);
% fbar = (1-t) .* repmat(f0,Q+2,1) + t .* repmat(f1,Q+2,1);
% wU0 = [reshape(mbar,(N+2)*(Q+1),1);reshape(fbar,(N+1)*(Q+2),1)];
% m = zeros(Q+1,N+1);
% f = mini*ones(Q+1,N+1);
% wV0 = [reshape(m,(N+1)*(Q+1),1);reshape(f,(N+1)*(Q+1),1)];
% minimumF0 = min(f0);
% % fin test % 

[XX,YY] = meshgrid(linspace(0,1,N+1),linspace(0,1,Q+1)); YY = flipud(YY);

% Itérations
niter = 3000;
cout = zeros(1,niter);
div = zeros(1,niter);
minF = zeros(1,niter);
tic;
for l = 1:niter
    [wU1 , wV1] = proxG1(2*zU0-wU0,2*zV0-wV0);
    wU1 = wU0 + alpha*(wU1- zU0); wV1 = wV0 + alpha*(wV1- zV0);
    
    [zU0,zV0] = proxG2(wU1,wV1);
    wU0 = wU1;
    wV0 = wV1;
    
    if mod(l,20) == 0
        f = reshape(zV0((N+1)*(Q+1)+1:end),Q+1,N+1);
        surf(XX,YY,f)
        xlabel('x');
        ylabel('t');
        zlabel('f');
        title(['itération : ',num2str(l)]);
        drawnow
       % pause
    end
    
    cout(l) = cost(zV0);
    div(l)  = sum(D*zU0);
    minF(l) = min(zV0((N+1)*(Q+1)+1:end));
end
toc

figure;
subplot(3,1,1)
plot([1:niter],cout);
title('cout')
subplot(3,1,2)
plot([1:niter],div)
title('div')
subplot(3,1,3)
plot([1:niter],minF);
title('min de f');

File: src/prototype/Fortran/transport.f90
program transport
implicit none

include 'global.inc'

integer :: i,j,k;
double precision :: mu = 0.1 , sigma = 0.005 , minimum = 0.000001;

double precision, dimension(N+1) :: f0, f1;

! projection sur C !
double precision, dimension(2*Q+2*N+4+(N+1)*(Q+1),2*(Q+1)*(N+1)+Q+N+2) :: A, Atmp;
double precision, dimension(2*Q+2*N+4+(N+1)*(Q+1),2*Q+2*N+4+(N+1)*(Q+1)) :: delta;
double precision, dimension((N+1)*(Q+1)+2*(Q+1)+2*(N+1)) :: y;
double precision, dimension(2*(N+1)*(Q+1)+N+Q+2,2*(N+1)*(Q+1)+N+Q+2) :: IdPC;
double precision, dimension(2*(N+1)) :: tmpPC;

! Proximal of G2 !
double precision, dimension((N+1)*(Q+2)+(N+2)*(Q+1),(N+1)*(Q+2)+(N+2)*(Q+1)) :: tmpPG2,tmptmpPG2;

! Variables DR ! 
double precision, dimension((N+2)*(Q+1)+(N+1)*(Q+2)) :: wU0,wU1,zU;
double precision, dimension(2*(N+1)*(Q+1)) :: wV0,wV1,zV;

! variables export files 
character(10) :: charI;
character(200) :: cmd;
double precision, dimension(Q+1,N+1) :: tmpZV;

! variables construction des inverses !
integer INFO;
integer, dimension((N+1)*(Q+1)+2*(Q+1)+2*(N+1)) :: IPIVy;
integer, dimension(2*(Q+1)*(N+1)+Q+N+2) :: IPIVA;
integer, dimension((N+1)*(Q+2)+(N+2)*(Q+1)) :: IPIV;
double precision, dimension((N+1)*(Q+2)+(N+2)*(Q+1)) :: work;

external SGESV
external DGETRF
external DGETRI

!!! Construction des matrices opérateurs !!!

print *, 'Begin Initialisation :'

call gauss(mu,sigma,N,f0);
call gauss(mu+0.8,sigma,N,f1);

f0 = f0 + minimum;
f1 = f1 + minimum;

f0 = f0/sum(f0(:));
f1 = f1/sum(f1(:));

print *, 'Construction of matrix operators'

call boundary(B);
call interpolation(Interp);
call divergence(D);

print *, 'Construction of the projection on C'
! projection sur C ! 
A(1:(N+1)*(Q+1),1:2*(Q+1)*(N+1)+Q+N+2) = D;
A((N+1)*(Q+1)+1:2*Q+2*N+4+(N+1)*(Q+1),1:2*(Q+1)*(N+1)+Q+N+2) = B;
delta = matmul(A,transpose(A));

y(1:(N+1)*(Q+1)) = 0; ! condition de divergence
y((N+1)*(Q+1)+1:(N+1)*(Q+1)+2*(Q+1)) = 0; ! condition sur m
do i = 1,2*(N+1)
	if (modulo(i,2) .EQ. 1) then
		tmpPC(i) = f1((i+1)/2); ! on met f1
	else
		tmpPC(i) = f0(i/2); ! on met f0
	end if
end do
y((N+1)*(Q+1)+2*(Q+1)+1:(N+1)*(Q+1)+2*(Q+1)+2*(N+1)) = tmpPC;

do i = 1,2*Q+2*N+4+(N+1)*(Q+1)
	print *, delta(i,:), 'ENDL';
end do

!call SGESV(2*Q+2*N+4+(N+1)*(Q+1),(N+1)*(Q+1)+2*(Q+1)+2*(N+1),delta,2*Q+2*N+4+(N+1)*(Q+1),IPIVy,y,2*Q+2*N+4+(N+1)*(Q+1),INFO);
call SGESV(2*Q+2*N+4+(N+1)*(Q+1),1,delta,2*Q+2*N+4+(N+1)*(Q+1),IPIVy,y,(N+1)*(Q+1)+2*(Q+1)+2*(N+1),INFO);

Cst = matmul(transpose(A),y);

Atmp = A;
call SGESV(2*Q+2*N+4+(N+1)*(Q+1),2*(Q+1)*(N+1)+Q+N+2,delta,2*Q+2*N+4+(N+1)*(Q+1),IPIVA,A,2*Q+2*N+4+(N+1)*(Q+1),INFO);

IdPC = 0;
do i = 1,(N+1)*(Q+2)+(N+2)*(Q+1)
	IdPC(i,i) = 1;
end do
P = IdPC - matmul(transpose(Atmp),A);

print *, 'Construction of the projection on G2'

! projection sur G2 !
tmpPG2 = 0;
do i = 1,(N+1)*(Q+2)+(N+2)*(Q+1)
	tmpPG2(i,i) = 1;
end do 
tmptmpPG2 = tmpPG2 + matmul(transpose(Interp),Interp);
call DGETRF((N+1)*(Q+2)+(N+2)*(Q+1),(N+1)*(Q+2)+(N+2)*(Q+1),tmptmpPG2,(N+1)*(Q+2)+(N+2)*(Q+1),IPIV,INFO);
call DGETRI((N+1)*(Q+2)+(N+2)*(Q+1),tmptmpPG2,(N+1)*(Q+2)+(N+2)*(Q+1),IPIV,work,(N+1)*(Q+2)+(N+2)*(Q+1),INFO);
pG2 = tmptmpPG2;

!! CHECK !!

print *, 'tmpC :';
do i = 1,2*(N+1)*(Q+1)+N+Q+2
!print *, Cst(i) , 'ENDL';
end do
!call check(INFO);

stop 0
!! END CHECK !!

! Début ! 
print *, 'Begin DR algorithm'
wU0 = 0; wV0 = 0;
wU1 = 0; wV1 = 0;
zU = 0; zV = 0;

do i = 1,niter
	call proxG1(2*zU-wU0,2*zV-wV0,wU1,wV1);
	wU1 = wU0 + alpha*(wU1 - zU);
	wV1 = wV0 + alpha*(wV1 - zV);
	
	call proxG2(wU1,wV1,zU,zV);
	
	wU0 = wU1;
	wV0 = wV1;
	
	write(charI,'(I5.5)') i
	open(7,file='results/f_'//trim(charI)//'.dat'); 
	write(7,*) "# ", "X ", "T ", "Z";
	tmpZV = reshape(zV,(/Q+1,N+1/));
	do j = 1,N+1 
		do k = 1,Q+1 
			write(7,*) (j-1)/(1.0*N),(k-1)/(1.0*Q), tmpZV(k,j); 
		end do 
	end do
	close(7);
	write(cmd,*) 'set contour; set dgrid3d ', N, ',', Q, ';', 'splot "results/f_'//trim(charI), &
		'.dat" with lines; set title "Iteration Nb '//trim(charI),' "';
	open(8,file="plot.gnu"); write(8,*) cmd; close(8);
	call system("gnuplot -p plot.gnu");

end do


















end program
